<a href="https://colab.research.google.com/github/farahh001/AI_Assignments/blob/main/Farahs'_Assignment_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Part 1: Load the dataset

In [109]:
# Load the dataset (load remotely, not locally)
import numpy as np
import matplotlib.pyplot as plt

import pandas as pd
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
# Measuring Accuracy
from sklearn import metrics
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier



In [110]:
# Output the first 15 rows of the data
iris_data = load_iris()
df = pd.DataFrame(iris_data['data'], columns = iris_data.feature_names) 
df["species"] = iris_data.target # adding species column


In [111]:
#print top 15 of the dataset
df.head(15)


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    species  
0         0  
1         0  
2         0  
3         0  
4         0  
5         0  
6         0  
7         0  
8         0  
9         0  
10        0  
11        0  
12        0  
13        0  
14        0

In [112]:
# Display a summary of the table information (number of datapoints, etc.)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   species            150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


In [113]:
df['species'].unique()

array([0, 1, 2])

In [114]:
#some statistical information
df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)     species  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

**Answer**

The datset contains five columns such as Petal Length, Petal Width, Sepal Length, Sepal Width, and Species. Three Species are found namely setosa,virginica and versicolor. Iris flowers can be categorized from sepal_length,	sepal_width,petal_length,petal_width and which they belong to species can be recorded 
 
 

# Part 2: Split the dataset into train and test

In [115]:
# Take the dataset and split it into our features (X) and label (y)
#  the following will select the first 4 columns except the last column
X = df.iloc[:, :-1].values


# here we are selecting the last column which is our dependent feature.
y = df.iloc[:,-1].values




In [116]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, train_size=0.9, random_state=0)
print ('The Training Set: %d rows\nTest Set: %d rows' % (X_train.shape[0], X_test.shape[0]))

The Training Set: 135 rows
Test Set: 15 rows


In [117]:
X

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [118]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

# Part 3: Logistic Regression

In [185]:
# i. Use sklearn to train a LogisticRegression model on the training set
classifier = LogisticRegression(max_iter =500,random_state = 42)
classifier.fit(X_train, y_train)

LogisticRegression(max_iter=500, random_state=42)

In [178]:
# ii. For a sample datapoint, predict the probabilities for each possible class

In [179]:
print (classifier.predict([[7.0,3.2,4.7,1.4]]))

[1]


In [180]:
print (classifier.predict([[4.9,3.0,1.4,0.2]]))

[0]


In [181]:
print (classifier.predict([[6.3,3.3,6.0,2.5]]))

[2]


In [182]:
#probability of each species
pd.DataFrame(classifier.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

Setosa  Versicolor  Virginica
0  0.052595    0.017443   0.929961

In [183]:
# getting the Training Set Predictions
y_train_pred = classifier.predict(X_train)

# getting the Test Set Predictions
y_test_pred = classifier.predict(X_test)

In [184]:
# Predicting the Test set results
print('The  test  accuracy of the Logistic Regression is:',metrics.accuracy_score(y_test_pred,y_test))
print('training Accuracy of the Model: ', metrics.accuracy_score(y_train, y_train_pred))

The  test  accuracy of the Logistic Regression is: 1.0
training Accuracy of the Model:  0.9703703703703703


It was a very small dataset and it seems like the logistic regression model performed greatly on this dataset.

In [127]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
classifier.coef_

array([[-0.41087786,  0.92630746, -2.43464502, -1.04770092],
       [ 0.40315462, -0.32623036, -0.11363406, -0.88710526],
       [ 0.00772324, -0.6000771 ,  2.54827907,  1.93480618]])

In [128]:
classifier.intercept_

array([  9.46338452,   2.44534908, -11.9087336 ])

# Part 4: Support Vector Machine

In [129]:
# i. Use sklearn to train a Support Vector Classifier on the training set
classifier_svm = SVC(kernel = 'linear',random_state= 0,probability=True)
classifier_svm.fit(X, y)


SVC(kernel='linear', probability=True, random_state=0)

In [130]:
# ii. For a sample datapoint, predict the probabilities for each possible class
print (classifier_svm.predict([[7.0,3.2,4.7,1.4]])) # 1 for Versicolor


[1]


In [131]:
print (classifier_svm.predict([[4.9,3.0,1.4,0.2]])) #0 for Setosa

[0]


In [132]:
print (classifier_svm.predict([[6.3,3.3,6.0,2.5]])) # 2 for Virginica

[2]


In [133]:
#probability of each species
pd.DataFrame(classifier_svm.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])

Setosa  Versicolor  Virginica
0  0.042563    0.003766   0.953671

In [155]:
#predicting the results
prediction_svm = classifier_svm.predict(X_test)
print(np.concatenate((prediction_svm.reshape(len(prediction_svm),1),y_test.reshape(len(y_test),1)),1))

[[2 2]
 [1 1]
 [0 0]
 [2 2]
 [0 0]
 [2 2]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [2 2]
 [1 1]
 [1 1]
 [1 1]
 [1 1]]


We are comparing the predict decision to the real ones, on the left side we have predicted test set result and on the right we have actual Species result. we can see that almost all the species are correctly categorized by our model


In [135]:
# iii. Report on the score for the SVM, what does the score measure?
print('The  test  accuracy of the SVM model is:',metrics.accuracy_score(prediction_svm,y_test))

The  test  accuracy of the SVM model is: 1.0


# Part 5: Neural Network

In [159]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
nn_model1= MLPClassifier(hidden_layer_sizes=(1),max_iter=2000)
nn_model1.fit(X_train, y_train)
print('The  test  accuracy of the Neural network(adam) model is:' ,nn_model1.score(X_test, y_test))

The  test  accuracy of the Neural network(adam) model is: 0.5333333333333333


It dropped quite a lot 53.33%. There are 4 input features and the 1 output feature. 
Since I don't have much data I started with 1 hidden layers and as you can see the model did poorly.

In [176]:
nn_model= MLPClassifier(hidden_layer_sizes=(7),max_iter=2000)
nn_model.fit(X_train, y_train)
print('The  test  accuracy of the Neural network(adam) model is:' ,nn_model.score(X_test, y_test))

The  test  accuracy of the Neural network(adam) model is: 1.0


as I increased the layers to 7, the model gave perfect score.

In [137]:
#2nd explore
nn_adam = MLPClassifier(hidden_layer_sizes=(150,100,50), max_iter=300,activation = 'relu',solver='adam',random_state=1)
nn_model2 = nn_adam.fit(X_train, y_train)



In [138]:
print (nn_model2.predict([[4.9,3.0,1.4,0.2]])) #0 for Setosa

[0]


In [139]:
print (nn_model2.predict([[7.0,3.2,4.7,1.4]])) # 1 for Versicolor

[1]


In [140]:
print (nn_model2.predict([[6.3,3.3,6.0,2.5]])) # 2 for Virginica

[2]


In [141]:
# ii. For a sample datapoint, predict the probabilities for each possible class
pd.DataFrame(nn_model2.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])


Setosa  Versicolor  Virginica
0  1.829486e-08    0.000028   0.999972

In [142]:
#predicting the results
prediction_nn2 = nn_adam.predict(X_test)


In [143]:
# iii. Report on the score for the Neural Network, what does the score measure?
print('The  test  accuracy of the Neural network(adam) model is:',metrics.accuracy_score(prediction_nn2,y_test))



The  test  accuracy of the Neural network(adam) model is: 1.0


In [144]:

# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)


In [145]:
#3rd explore 
nn_model3 = MLPClassifier(max_iter = 1500, solver='sgd', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1).fit(X_train, y_train)
print('The  test  accuracy of the Neural network(sgd) model is:', nn_model3.score(X_test, y_test))

The  test  accuracy of the Neural network(sgd) model is: 1.0


# Part 6: K-Nearest Neighbors

In [151]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
classifier_KNN = KNeighborsClassifier(n_neighbors=5).fit(X_train, y_train) 

In [154]:
# Note: KNN is a nonparametric model and technically doesn't require training
# predicting results 
pred_KNN= classifier_KNN.predict(X_test)
print(np.concatenate((pred_KNN.reshape(len(pred_KNN),1),y_test.reshape(len(y_test),1)),1))


[[2 2]
 [1 1]
 [0 0]
 [2 2]
 [0 0]
 [2 2]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [2 2]
 [1 1]
 [1 1]
 [1 1]
 [1 1]]


In [148]:
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier


In [153]:
# ii. For a sample datapoint, predict the probabilities for each possible class
pd.DataFrame(classifier_KNN.predict_proba([[1, 2, 3, 4]]), columns = ["Versicolor", "Virginica","Setosa"])

Versicolor  Virginica  Setosa
0         0.0        0.8     0.2

In [160]:
# iii. Report on the score for kNN, what does the score measure?
print('The  test  accuracy of the KNN model is:', classifier_KNN.score(X_test, y_test))

The  test  accuracy of the KNN model is: 1.0


# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

I played around with different hidden layers, then tried to see if it actually improve the performance. I have also tried on optimization algorithm essentially Adam which is an algorithm for gradient-based optimization of stochastic objective functions and SGD which is also a variant of gradient descent. The model showed consistent 100% accuracy. It came as a surprise to me to see perfect scores. The datset is fairly small hence the results were predicted correctly by the models. 
